### PR TITLE
Search return of ChannelSftp.ls() for an object with the specified filename instead of the filename itself.

### LINKED ISSUE #3784: bug on SftpConnection.exists( ) method
On 3.2.x and trunk

I suspect a bug on method SftpConnection.exists( )

public boolean exists(String file, String path) {
try {
cwd(path);

return client.ls(".").contains(file);

} catch (Exception e) {
return false;
}
}

The contains( ) does not work on LsEntry.

Should be fixed by
@Override
public boolean exists(String file, String path) {
try {
cwd(path);
@SuppressWarnings("unchecked")
Vector<ChannelSftp.LsEntry> entries = client.ls(".");

for (Iterator<ChannelSftp.LsEntry> iter = entries.iterator(); iter.hasNext() {
if ( iter.next().getFilename().compareTo(file) == 0)
return true;
}
return false;

} catch (Exception e) {
return false;
}
}

Imported Issue. Original Details:
Jira Issue Key: MIRTH-3903
Reporter: ptisserand
Created: 2016-03-11T02:00:34.000-0800

### PR DESCRIPTION
**Pull request by ChristopherSchultz**
This fixes #3784.

### CODE REVIEW

File: server/src/com/mirth/connect/connectors/file/filesystems/SftpConnection.java [hunk 1]
```diff
@@ -222,9 +222,18 @@ public List<String> listDirectories(String fromDir) throws Exception {
...
         try {
             cwd(path);
-            return client.ls(".").contains(file);
+
+            @SuppressWarnings("unchecked")
+            Vector<ChannelSftp.LsEntry> entries = (Vector<ChannelSftp.LsEntry>)client.ls(".");
+
+            return entries.stream()
+                    .anyMatch(le -> file.equals(le.getFilename()));
         } catch (Exception e) {
```
Comment (jonbartels) [inline server/src/com/mirth/connect/connectors/file/filesystems/SftpConnection.java:237]: Should you add a logging statement here?

If an exception was thrown and logged, it would give the user some information about what went wrong. 

Logging improvements might be out of scope though, should they be a separate PR?

File: server/src/com/mirth/connect/connectors/file/filesystems/SftpConnection.java [hunk 2]
```diff
@@ -222,9 +222,18 @@ public List<String> listDirectories(String fromDir) throws Exception {
 
     @Override
     public boolean exists(String file, String path) {
+        if(null == file) {
```
Comment (jonbartels) [inline server/src/com/mirth/connect/connectors/file/filesystems/SftpConnection.java:225]: Should this also check null for path?

File: server/src/com/mirth/connect/connectors/file/filesystems/SftpConnection.java (hunk 1 above)
Comment (ChristopherSchultz) [reply server/src/com/mirth/connect/connectors/file/filesystems/SftpConnection.java:237]: 💯 

I put this into the bug report but I didn't put it into my PR. I'll definitely add that.

File: server/src/com/mirth/connect/connectors/file/filesystems/SftpConnection.java (hunk 2 above)
Comment (ChristopherSchultz) [reply server/src/com/mirth/connect/connectors/file/filesystems/SftpConnection.java:225]: Good question. I checked the `file` for `null` to avoid any possibility of NPE in this exact method. I suppose checking `path` for `null` could work, too.

I'm thinking that if either are `null` maybe `IllegalArgumentException` is more appropriate than returning `false`. Comments?

Comment (pladesma) [reply server/src/com/mirth/connect/connectors/file/filesystems/SftpConnection.java:225]: Checking `path` for `null` wouldn't hurt, but I don't think it's strictly necessary in this case. Like you mentioned, checking `file` for `null` is mainly to avoid an NPE, which is not as issue with `path`.

There wouldn't  anything wrong with throwing an `IllegalArgumentException`, but since the existing behavior of the method is to catch all exceptions and return `false`, I lean toward just returning `false` to avoid any unforeseen bugs.

Summing up, I wouldn't make any changes here.

Comment (ChristopherSchultz) [reply server/src/com/mirth/connect/connectors/file/filesystems/SftpConnection.java:225]: @pladesma It's your product, so I'll do whatever you want, but I'd rather have an error thrown than for this check to return `false` and then have something very very weird happen later. I see "returning false" as more likely to cause unforeseen bugs rather than "throw exception".

I suppose "can't write file" is the likely next error you'll encounter if this method returns `false` because `path`/`file` are not sane.
